drivers/flash/flash_mcux_flexspi_nor.c

```diff
@@ -0,0 +1,515 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	nxp_imx_flexspi_nor
+
+#include <drivers/flash.h>
+#include <logging/log.h>
+#include <sys/util.h>
+#include "spi_nor.h"
+#include "flash_mcux_flexspi.h"
+
+#ifdef CONFIG_HAS_MCUX_CACHE
+#include <fsl_cache.h>
+#endif
+
+#define NOR_WRITE_SIZE	1
+#define NOR_ERASE_VALUE	0xff
+
+LOG_MODULE_DECLARE(flash_flexspi, CONFIG_FLASH_LOG_LEVEL);
+
+enum {
+	/* SPI instructions */
+	READ_ID,
+	READ_STATUS_REG,
+	WRITE_STATUS_REG,
+	WRITE_ENABLE,
+	PAGE_PROGRAM,
+	ERASE_SECTOR,
+	ERASE_CHIP,
+
+	/* Quad SPI instructions */
+	READ_FAST_QUAD_OUTPUT,
+	PAGE_PROGRAM_QUAD_INPUT,
+	ENTER_QPI,
+};
+
+struct flash_flexspi_nor_config {
+	char *controller_label;
+	flexspi_port_t port;
+	flexspi_device_config_t config;
+	struct flash_pages_layout layout;
+	struct flash_parameters flash_parameters;
+};
+
+struct flash_flexspi_nor_data {
+	const struct device *controller;
+};
+
+static const uint32_t flash_flexspi_nor_lut[][4] = {
+	[READ_ID] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, SPI_NOR_CMD_RDID,
+				kFLEXSPI_Command_READ_SDR,  kFLEXSPI_1PAD, 0x04),
+	},
+
+	[READ_STATUS_REG] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, SPI_NOR_CMD_RDSR,
+				kFLEXSPI_Command_READ_SDR,  kFLEXSPI_1PAD, 0x04),
+	},
+
+	[WRITE_STATUS_REG] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, SPI_NOR_CMD_WRSR,
+				kFLEXSPI_Command_WRITE_SDR, kFLEXSPI_1PAD, 0x04),
+	},
+
+	[WRITE_ENABLE] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, SPI_NOR_CMD_WREN,
+				kFLEXSPI_Command_STOP,      kFLEXSPI_1PAD, 0),
+	},
+
+	[PAGE_PROGRAM] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, SPI_NOR_CMD_PP,
+				kFLEXSPI_Command_RADDR_SDR, kFLEXSPI_1PAD, 0x18),
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_WRITE_SDR, kFLEXSPI_1PAD, 0x04,
+				kFLEXSPI_Command_STOP,      kFLEXSPI_1PAD, 0),
+	},
+
+	[ERASE_SECTOR] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, 0xD7,
```

Comment:
Why is command `D7h` used here, not `20h` (`SPI_NOR_CMD_SE`)? W25Q128JV as well as IS25WP064A support command `20h`, but W25Q128JV not support `D7h`. I found command `20h` in JESD251A (table 5), but i cannot understand is this command standard.

Reply:
This is an interesting point.  The command might be defaultable, but for portability it should come from the JESD216 BFP DW8 and DW9.  Is it possible to interact with the device using a default LUT first, to extract the SFDP data, then program a LUT that is correct for the discovered device?

Reply:
The LUT is derived from the MCUXpresso SDK example `evkmimxrt1064_flexspi_nor_polling_transfer`. I don't know why `0xD7` was used there instead of `0x20`, but experimentally it seems that the standard command also works with IS25WP064.

Reply:
How can different flash chips be supported in the future? I guess the LUT might be subject to change if another chip is used, but now it is hardcoded in the driver.

Reply:
For exapmle, board [VisionSOM-RT](https://wiki.somlabs.com/index.php/VisionSOM-RT) have flash [W25Q128JV](https://www.winbond.com/resource-files/w25q128jv%20revf%2003272018%20plus.pdf), which do not support `D7h`. Also [S25FL128L](https://www.cypress.com/file/316171/download) by CYPRESS and [MT25QL128](https://www.micron.com/products/nor-flash/serial-nor-flash/part-catalog?Value=128Mb&Id={826DD0CE-31A2-43CD-AA36-E65D0DB21D31}) by Micron support command `20h`. So, it is universal to use command `20h`.

Reply:
One of the main purposes of adding SFDP support in Zephyr is to allow the driver to get the information (like erase command) from the flash chip directly, and configure itself.  I do think doing that would handle the concerns in this thread, but other than changing the command to the standard 20h from the unusual D7h I don't think any of that has to be done now.